src/ast/passes/macro_expansion.cpp

```diff
@@ -0,0 +1,320 @@
+#include <unordered_map>
+#include <unordered_set>
+#include <utility>
+
+#include "ast/ast.h"
+#include "ast/context.h"
+#include "ast/passes/macro_expansion.h"
+#include "ast/visitor.h"
+#include "bpftrace.h"
+
+#include "log.h"
+
+namespace bpftrace::ast {
+
+class MacroExpander : public Visitor<MacroExpander> {
+public:
+  MacroExpander(ASTContext &ast,
+                std::string macro_name,
+                const std::unordered_map<std::string, Macro *> &macros,
+                std::vector<std::string> &&macro_stack);
+
+  using Visitor<MacroExpander>::visit;
+
+  void visit(AssignVarStatement &assignment);
+  void visit(Variable &var);
+  void visit(Map &map);
+  void visit(Expression &expr);
+
+  std::optional<Block *> expand(Macro &macro, const Call &call);
+  bool check_recursive_call(const std::string &macro_name, const Call &call);
+
+private:
+  ASTContext &ast_;
+  const std::string macro_name_;
+  const std::unordered_map<std::string, Macro *> &macros_;
+
+  std::string get_new_var_prefix();
+  std::string get_new_var_ident(std::string original_ident);
+
+  // Maps of macro map/var names -> callsite map/var names
+  std::unordered_map<std::string, std::string> maps_;
+  std::unordered_map<std::string, std::string> vars_;
+  std::unordered_set<std::string> temp_vars_;
+  const std::vector<std::string> macro_stack_;
+};
+
+class MacroCollection : public Visitor<MacroCollection> {
+public:
+  MacroCollection(ASTContext &ast, BPFtrace &b);
+  void collect_macros();
+
+  using Visitor<MacroCollection>::visit;
+
+  void visit(Expression &expr);
+
+private:
+  ASTContext &ast_;
+  BPFtrace &bpftrace_;
+  std::unordered_map<std::string, Macro *> macros_;
+};
+
+MacroExpander::MacroExpander(
+    ASTContext &ast,
+    std::string macro_name,
+    const std::unordered_map<std::string, Macro *> &macros,
+    std::vector<std::string> &&macro_stack)
+    : ast_(ast),
+      macro_name_(std::move(macro_name)),
+      macros_(macros),
+      macro_stack_(std::move(macro_stack))
+{
+  assert(!macro_stack_.empty());
+}
+
+void MacroExpander::visit(AssignVarStatement &assignment)
+{
+  auto *var = assignment.var();
+  if (var->ident.starts_with(get_new_var_prefix())) {
```

Comment:
Doing a variable name check seems fragile and leaky. Can you store a std::unordered_set of pointers to variables that this pass created and check that instead?